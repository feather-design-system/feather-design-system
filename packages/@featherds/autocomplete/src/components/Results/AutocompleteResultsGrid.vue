<template>
  <div class="feather-autocomplete-results-grid" ref="grid">
    <table
      class="feather-autocomplete-results-grid-container"
      :class="cls"
      tabindex="-1"
      aria-hidden="false"
      aria-live="polite"
      role="grid"
      :aria-multiselectable="single ? 'false' : 'true'"
    >
      <thead>
        <tr role="row">
          <th v-for="item in config" :key="item.title">
            {{ item.title }}
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          role="row"
          v-for="(item, index) in items"
          :key="item[textProp]"
          :aria-selected="isSelected(item)"
          :class="{ focus: isActive(index), selected: isSelected(item) }"
          @click.stop="select(item)"
        >
          <td
            v-for="(col, colIndex) in config"
            :key="item[textProp] + col.prop"
            :id="getId(index, colIndex)"
            :class="{ 'focus-cell': isActiveCell(index, colIndex) }"
          >
            <Highlighter
              v-if="col.prop === textProp"
              :highlight="highlight"
              :query="query"
              :text="item[col.prop]"
            />
            <p v-else>{{ item[col.prop] }}</p>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</template>
<script>
import { toView } from "@featherds/utils/scroll";
import Highlighter from "../Highlight/Highlighter";
import HighlightMixin from "../Highlight/HighlightMixin";
import HighlighterMixin from "../Highlight/HighlighterMixin";
export default {
  mixins: [HighlightMixin, HighlighterMixin],
  emits: ["select"],
  props: {
    activeId: {
      type: String,
      required: true,
    },
    activeRow: {
      type: Number,
      required: true,
    },
    activeCol: {
      type: Number,
      required: true,
    },
    items: {
      type: Array,
      required: true,
    },
    value: {
      type: [Array, Object],
      default: () => [],
    },
    textProp: {
      type: String,
      default: "_text",
    },
    single: {
      type: Boolean,
      default: false,
    },
    config: {
      type: Array,
      required: true,
    },
  },
  watch: {
    activeRow(index) {
      if (index > -1) {
        this.$nextTick(() => {
          const el = Array.prototype.slice.call(
            this.$el.querySelectorAll("tr")
          )[index + 1]; //+1 for header
          toView(el, this.$refs.grid);
        });
      }
    },
  },
  computed: {
    cls() {
      return this.config.map((item, index) => {
        if (item.align && item.align.toLowerCase() === "right") {
          return `tr${index + 1}`;
        }
        if (item.align && item.align.toLowerCase() === "center") {
          return `tc${index + 1}`;
        }
        return `tl${index + 1}`;
      });
    },
  },
  methods: {
    isSelected(item) {
      if (this.value && this.value.length) {
        return this.value.some((x) => x[this.textProp] === item[this.textProp]);
      }
      return this.value[this.textProp] === item[this.textProp];
    },
    isActive(index) {
      return this.activeRow === index;
    },
    isActiveCell(row, col) {
      return this.activeRow === row && this.activeCol === col;
    },
    getId(index, col) {
      return index === this.activeRow && this.activeCol === col
        ? this.activeId
        : null;
    },
    select(item) {
      this.$emit("select", item);
    },
  },
  components: {
    Highlighter,
  },
};
</script>
<style lang="scss" scoped>
@import "~@featherds/styles/mixins/typography";
@import "~@featherds/styles/themes/variables";
@import "~@featherds/styles/themes/utils";
@import "~@featherds/table/scss/_table";
.feather-autocomplete-results-grid {
  overflow-y: scroll;
}
.feather-autocomplete-results-grid-container {
  @include table();
  @include table-condensed();
  @include row-hover();
  @include row-select();
  width: 100%;
  tr {
    &.selected {
      color: var($primary-text-on-surface);
    }

    &.focus td:first-child {
      box-shadow: inset 3px 0 0px 0px var($primary);
    }
    td {
      border: 1px solid transparent;
    }
    td.focus-cell {
      border: 1px solid var($primary);
    }
  }
  tbody tr {
    cursor: pointer;
  }
}
</style>
<style lang="scss">
@import "../../../scss/mixins";
.feather-autocomplete-results-grid {
  @include autocomplete-results-height(6);
}
</style>
