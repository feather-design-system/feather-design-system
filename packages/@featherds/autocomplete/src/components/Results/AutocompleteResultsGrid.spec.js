import AutocompleteResultsGrid from "./AutocompleteResultsGrid.vue";
import { shallowMount, config as testUtilsConfig } from "@vue/test-utils";

testUtilsConfig.renderStubDefaultSlot = true;

const activeId = "ACTIVE";
const items = [
  {
    text: "Test1",
    number: 1,
  },
  {
    text: "Test2",
    number: 2,
  },
  {
    text: "Test3",
    number: 3,
  },
  {
    text: "Test4",
    number: 4,
  },
];
const config = [
  { title: "TEXT", prop: "text" },
  { title: "NUMBER", prop: "number", align: "right" },
];
const getProps = (props) => {
  return {
    activeId,
    config,
    query: "t",
    items: items.slice(0),
    ...props,
  };
};

const getWrapper = function (options = {}) {
  options.props = getProps(options.props || {});
  return shallowMount(AutocompleteResultsGrid, options);
};

describe("Autocomplete Results Grid", () => {
  it("should support single select", () => {
    const wrapper = getWrapper({
      props: {
        single: true,
        activeRow: -1,
        activeCol: -1,
      },
    });
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
  it("should support no item active", () => {
    const wrapper = getWrapper({
      props: {
        activeRow: -1,
        activeCol: -1,
      },
    });
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
  it("should support active index being out of range", () => {
    const wrapper = getWrapper({
      props: {
        activeRow: items.length + 1,
        activeCol: -1,
      },
    });
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
  it("should support active index within results range", () => {
    const wrapper = getWrapper({
      props: {
        activeRow: 0,
        activeCol: 0,
      },
    });
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
  it("should mark items as selected that appear in value array", () => {
    const wrapper = getWrapper({
      props: {
        activeRow: -1,
        activeCol: -1,
        value: [items[0]],
      },
    });
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
  it("should mark items as selected that appear in value array and active when active index is set.", () => {
    const wrapper = getWrapper({
      props: {
        activeRow: 0,
        activeCol: 0,
        value: [items[0]],
      },
    });
    expect(wrapper.wrapperElement).toMatchSnapshot();
  });
});
