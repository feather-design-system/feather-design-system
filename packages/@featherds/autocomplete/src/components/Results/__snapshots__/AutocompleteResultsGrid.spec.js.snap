// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Results Grid should mark items as selected that appear in value array 1`] = `
<div
  class="feather-autocomplete-results-grid"
>
  <table
    aria-hidden="false"
    aria-live="polite"
    aria-multiselectable="true"
    class="feather-autocomplete-results-grid-container tl1 tr2"
    role="grid"
    tabindex="-1"
  >
    <thead>
      <tr
        role="row"
      >
        
        <th>
          TEXT
        </th>
        <th>
          NUMBER
        </th>
        
      </tr>
    </thead>
    <tbody>
      
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test1
          </p>
        </td>
        <td
          class=""
        >
          <p>
            1
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test2
          </p>
        </td>
        <td
          class=""
        >
          <p>
            2
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test3
          </p>
        </td>
        <td
          class=""
        >
          <p>
            3
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test4
          </p>
        </td>
        <td
          class=""
        >
          <p>
            4
          </p>
        </td>
        
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`Autocomplete Results Grid should mark items as selected that appear in value array and active when active index is set. 1`] = `
<div
  class="feather-autocomplete-results-grid"
>
  <table
    aria-hidden="false"
    aria-live="polite"
    aria-multiselectable="true"
    class="feather-autocomplete-results-grid-container tl1 tr2"
    role="grid"
    tabindex="-1"
  >
    <thead>
      <tr
        role="row"
      >
        
        <th>
          TEXT
        </th>
        <th>
          NUMBER
        </th>
        
      </tr>
    </thead>
    <tbody>
      
      <tr
        aria-selected="true"
        class="focus selected"
        role="row"
      >
        
        <td
          class="focus-cell"
          id="ACTIVE"
        >
          <p>
            Test1
          </p>
        </td>
        <td
          class=""
        >
          <p>
            1
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test2
          </p>
        </td>
        <td
          class=""
        >
          <p>
            2
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test3
          </p>
        </td>
        <td
          class=""
        >
          <p>
            3
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test4
          </p>
        </td>
        <td
          class=""
        >
          <p>
            4
          </p>
        </td>
        
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`Autocomplete Results Grid should support active index being out of range 1`] = `
<div
  class="feather-autocomplete-results-grid"
>
  <table
    aria-hidden="false"
    aria-live="polite"
    aria-multiselectable="true"
    class="feather-autocomplete-results-grid-container tl1 tr2"
    role="grid"
    tabindex="-1"
  >
    <thead>
      <tr
        role="row"
      >
        
        <th>
          TEXT
        </th>
        <th>
          NUMBER
        </th>
        
      </tr>
    </thead>
    <tbody>
      
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test1
          </p>
        </td>
        <td
          class=""
        >
          <p>
            1
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test2
          </p>
        </td>
        <td
          class=""
        >
          <p>
            2
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test3
          </p>
        </td>
        <td
          class=""
        >
          <p>
            3
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test4
          </p>
        </td>
        <td
          class=""
        >
          <p>
            4
          </p>
        </td>
        
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`Autocomplete Results Grid should support active index within results range 1`] = `
<div
  class="feather-autocomplete-results-grid"
>
  <table
    aria-hidden="false"
    aria-live="polite"
    aria-multiselectable="true"
    class="feather-autocomplete-results-grid-container tl1 tr2"
    role="grid"
    tabindex="-1"
  >
    <thead>
      <tr
        role="row"
      >
        
        <th>
          TEXT
        </th>
        <th>
          NUMBER
        </th>
        
      </tr>
    </thead>
    <tbody>
      
      <tr
        aria-selected="true"
        class="focus selected"
        role="row"
      >
        
        <td
          class="focus-cell"
          id="ACTIVE"
        >
          <p>
            Test1
          </p>
        </td>
        <td
          class=""
        >
          <p>
            1
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test2
          </p>
        </td>
        <td
          class=""
        >
          <p>
            2
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test3
          </p>
        </td>
        <td
          class=""
        >
          <p>
            3
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test4
          </p>
        </td>
        <td
          class=""
        >
          <p>
            4
          </p>
        </td>
        
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`Autocomplete Results Grid should support no item active 1`] = `
<div
  class="feather-autocomplete-results-grid"
>
  <table
    aria-hidden="false"
    aria-live="polite"
    aria-multiselectable="true"
    class="feather-autocomplete-results-grid-container tl1 tr2"
    role="grid"
    tabindex="-1"
  >
    <thead>
      <tr
        role="row"
      >
        
        <th>
          TEXT
        </th>
        <th>
          NUMBER
        </th>
        
      </tr>
    </thead>
    <tbody>
      
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test1
          </p>
        </td>
        <td
          class=""
        >
          <p>
            1
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test2
          </p>
        </td>
        <td
          class=""
        >
          <p>
            2
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test3
          </p>
        </td>
        <td
          class=""
        >
          <p>
            3
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test4
          </p>
        </td>
        <td
          class=""
        >
          <p>
            4
          </p>
        </td>
        
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`Autocomplete Results Grid should support single select 1`] = `
<div
  class="feather-autocomplete-results-grid"
>
  <table
    aria-hidden="false"
    aria-live="polite"
    aria-multiselectable="false"
    class="feather-autocomplete-results-grid-container tl1 tr2"
    role="grid"
    tabindex="-1"
  >
    <thead>
      <tr
        role="row"
      >
        
        <th>
          TEXT
        </th>
        <th>
          NUMBER
        </th>
        
      </tr>
    </thead>
    <tbody>
      
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test1
          </p>
        </td>
        <td
          class=""
        >
          <p>
            1
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test2
          </p>
        </td>
        <td
          class=""
        >
          <p>
            2
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test3
          </p>
        </td>
        <td
          class=""
        >
          <p>
            3
          </p>
        </td>
        
      </tr>
      <tr
        aria-selected="true"
        class="selected"
        role="row"
      >
        
        <td
          class=""
        >
          <p>
            Test4
          </p>
        </td>
        <td
          class=""
        >
          <p>
            4
          </p>
        </td>
        
      </tr>
      
    </tbody>
  </table>
</div>
`;
